tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Analyzer.scala

```diff
@@ -349,19 +352,34 @@ final class Analyzer(semantics: Semantics, outputMode: OutputMode,
 
     def instantiated()(implicit from: From): Unit = {
       instantiatedFrom ::= from
-      if (!isInstantiated && isClass) {
+      if (!isInstantiated && (isScalaClass || isAnyRawJSType)) {
         isInstantiated = true
-        accessData()
-        ancestors.foreach(_.subclassInstantiated())
 
-        for ((methodName, from) <- delayedCalls)
-          delayedCallMethod(methodName)(from)
+        if (isScalaClass) {
+          accessData()
+          ancestors.foreach(_.subclassInstantiated())
+
+          for ((methodName, from) <- delayedCalls)
+            delayedCallMethod(methodName)(from)
+        } else {
+          assert(isAnyRawJSType)
+
+          subclassInstantiated()
+
+          if (isJSClass)
+            superClass.instantiated()
+
+          for (methodInfo <- methodInfos.values) {
+            if (methodInfo.isExported)
+              methodInfo.reach(this)(FromExports)
+          }
```

Comment:
Exports on normal classes take overriding into account. With:
```scala
class Foo {
  @JSExport def foobar(): Int = 42
}
class Bar extends Foo {
  @JSExport override def foobar(): Int = 3
}
new Bar
```
you'll have that `Foo["foobar"]` can be dce'ed away, because `Foo` is not instantiated. Only `Bar["foobar"]` is reachable. This is implemented through `callMethod` which actually *delays* the actual call until the class is `instantiated()`. We are allowed to do this because the semantics of Scala classes and exports do not guarantee anything about what will be present in the `prototype`s.

For JS classes, however, it is necessary to expose everything as soon as the constructor is reachable, because the classes could be extended by JS classes, or otherwise reflected upon using the JS prototype.